# 판다스

판다스는 파이썬에서 데이터처리를 위해 존재하는 가장 인기 있는 라이브러리<br>
일반적으로 데이터 세트는 2차원 데이터 (행_row X 열_column) -> 판다스는 2차원 데이터를 효율적으로 가공/처리할 수 있음

판다스의 핵심 객체<br> 
DataFrame: 여러 개의 행과 열로 이루어지는 2차원 데이터를 담는 데이터 구조체<br>
Series: 칼럼이 하나뿐인 데이터 구조체 (반면, DataFrame은 칼럼이 여러 개인 데이터 구조체)<br>
Index: 개별 데이터를 고유하게 식별하는 Key값<br>
=> Series와 DataFrame은 모두 Index를 key 값으로 가지고 있음<br>


## 판다스 시작 - 파일을 DataFrame으로 로딩, 기본 API

판다스는 다양한 포맷으로 된 파일을 DataFrame으로 로딩할 수 있는 편리한 API를 제공

read_csv()<br>
: CSV(칼럼을 ','로 구분한 파일 포맷)파일 포맷변환을 위한 API<br>
: 사실 CSV뿐만 아니라 어떤 필드 구분 문자 기반의 파일 포맷도 DataFrame으로 변환이 가능<br>
    -> 인자인 sep에 해당 구분 분자를 입력하면 됨 ex. read_csv("파일명",sep="\t")
    
read_table()<br>
: 디폴트 필드 구분 문자는 탭 문자 (read_csv()와의 가장 큰 차이는 필드 구분 문자가 콤마냐, 탭이냐)
    
read_fwf()<br>
: Fixed Width, 즉 고정 길이 기반의 칼럼 포맷을 DataFrame으로 로딩하기 위한 API

In [2]:
import pandas as pd

In [14]:
#워킹디렉터리가 아닐 경우

titanic_df = pd.read_csv("/Users/ijiseon/Desktop/titanic_train.csv")
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [5]:
#워킹디렉터리일 경우 

titanic_df = pd.read_csv("titanic_train.csv")
print("titanic 변수 type:", type(titanic_df))
titanic_df

titanic 변수 type: <class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

#### pd.read_csv()

파일명 인자로 들어온 파일을 로딩해 DataFrame 객체로 반환

반환된 DataFrame 객체<br>
- 데이터 파일의 첫번째 줄에 있던 칼럼 문자열이 DataFrame의 칼럼으로 할당<br>
= read_csv()는 별다른 파라미터 지정이 없으면 **파일의 맨 처음 로우를 칼럼명으로 인지**<br>
- 맨 왼쪽 파일에 기재돼 있지 않는 데이터값이 로우 순으로 0,1,2,3 ... 같이 순차적으로 표시<br>
=> 칼럼명이 표시되지 않는 이 데이터드리 바로 판다스의 Index 객체 값<br>
   모든 DataFrame 내의 데이터는 생성되는 순간 고유의 Index값을 가지게 됨.<br>

In [6]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [7]:
print("DataFrame 크기: ", titanic_df.shape)

DataFrame 크기:  (891, 12)


#### 메서드 info( ) 와 descirbe( )

In [8]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


#### info( )
- RangeIndex를 통해 전체 row 수를 알 수 있다. 또한 total columns 수 역시 알 수 있다<br>
- 칼럼별 데이터 타입<br>
; Name 칼럼의 데이터 타입은 object인데, 이는 문자열 타입으로 생각해도 무방<br> 
- non-null(Null값이 아님)값의 개수<br>
; Age 칼럼 891개 데이터 중 714개가 Null이 아님 = 177개가 Null<br>

In [9]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

#### describe( )
- 오직 숫자형(int,float 등) 칼럼의 분포도만 조사하여 자동으로 object 타입의 칼럼은 출력에서 제외
- count는 not null인 데이터 건수
- **해당 숫자 칼럼이 숫자형 카테고리 칼럼인지를 판단할 수 있게 도와준다**<br>
; 카테고리 칼럼은 특정 범주에 속하는 값을 코드화한 칼럼 ex) '남'을 1, '여'를 2

### 메서드 value_counts

step1. DataFrame의 [ ] 연산자 내부에 칼럼명을 입력<br>
step2. Series 형태로 특정 칼럼 데이터 세트가 반환<br>
step3. 반환된 Series 객체에 value_counts( ) 메서드를 호출

=> 해당 칼럼값의 유형과 건수를 확인할 수 있음<br>
value_counts( )는 많은 건수 순서로 정렬되어 값을 반환<br>
**DataFrame은 value_counts()메서드를 가지고 있지 않음**<br>

In [10]:
value_counts = titanic_df["Pclass"].value_counts()
print(value_counts)

3    491
1    216
2    184
Name: Pclass, dtype: int64


In [11]:
titanic_pclass = titanic_df["Pclass"]
print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


In [12]:
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

In [13]:
value_counts = titanic_df["Pclass"].value_counts()
print(type(value_counts))
print(value_counts)

<class 'pandas.core.series.Series'>
3    491
1    216
2    184
Name: Pclass, dtype: int64


type(titanic_pclass)와 type(value_counts) 모두 seires<br>
모두 왼쪽이 인덱스값, 오른쪽이 데이터값<br>
그런데 후자의 경우, 인덱스가 단순히 0부터 시작하는 순차 값이 아님<br>
; Pclass 칼럼 값이 3,1,2를 나타내고 있음<br>
=> 인덱스는 단순히 순차 값과 같은 의미 없는 식별자만 할당하는 것이 아니라, 고유성이 보장된다면 의미 있는 데이터값 할당도 가능<br>
value_counts()는 칼럼 값별 데이터 건수를 반환하므로, 고유 칼럼 값을 식별자로 사용할 수 있음

인덱스는 DataFrame, Series가 만들어진 후에도 변경 가능<br>
또한 인덱스는 숫자형뿐만 아니라 문자열도 가능<br>
단, 모든 인덱스는 고유성이 보장돼야 함<br>

----------------------------------------------------------------------------------------------------------

## DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호변환

## 넘파이 ndarray, 리스트, 딕셔너리 -> DataFrame

DataFrame은 리스트와 넘파이 ndarray와 다르게 칼럼명을 가지고 있습니다<br>
=> 일반적으로 DataFrame으로 변환 시, 이 칼럼명을 지정해 줍니다. (지정하지 않으면 자동으로 칼럼명 할당)

판다스 DataFrame 객체<br>
생성인자 data는 리스트, 딕셔너리, 넘파이 ndarray를 입력 받음<br>
생성인자 columns는 칼럼명 리스트를 입력 받음
                     
DataFrame은 기본적으로 행과 열을 가지는 2차원 데이터 = 2차원 이하의 데이터들만 DataFrame으로 변환 가능

In [14]:
import numpy as np

In [15]:
col_name1 = ["col1"]

list1 = [1,2,3]
array1 = np.array(list1)
print("array1 shape:", array1.shape)

array1 shape: (3,)


In [17]:
#리스트를 이용해 DataFrame 생성
df_list1 = pd.DataFrame(list1, columns=col_name1)
print("1차원 리스트로 만든 DataFrame:\n", df_list1)

1차원 리스트로 만든 DataFrame:
    col1
0     1
1     2
2     3


In [18]:
#넘파이를 이용해 DataFrame 생성
df_array1 = pd.DataFrame(array1, columns=col_name1)
print("1차원 ndarray로 만든 DataFrame:\n", df_array1)

1차원 ndarray로 만든 DataFrame:
    col1
0     1
1     2
2     3


In [20]:
#2행X3열 형태의 리스와 ndarray 생성한 뒤 이를 Data Frame으로 변환 = 3개의 칼럼명이 필요함

col_name2 = ["col1","col2","col3"]

list2 = [[1,2,3],
        [11,12,13]]

array2 = np.array(list2)

print("array2 shape:", array2.shape)
print("---------------------------")

df_list2 = pd.DataFrame(list2, columns=col_name2)
print("2차원 리스트로 만든 DataFrame:\n", df_list2)
print("---------------------------")

df_array2 = pd.DataFrame(array2, columns=col_name2)
print("2차원 ndarray로 만든 DataFrame:\n", df_array2)

array2 shape: (2, 3)
---------------------------
2차원 리스트로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13
---------------------------
2차원 ndarray로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13


딕셔너리를 DataFrame으로 변환<br>
: 딕셔너리의 키(Key)는 칼럼명으로, 딕셔너리의 값(Value)는 키에 해당하는 칼럼 데이터로 변환<br>
: 키의 경우는 문자열, 값의 경우 리스트(또는 ndarray) 형태로 딕셔너리를 구성

In [21]:
#key는 문자열 칼럼명으로 매핑, value는 리스트 형(또는 ndarray) 칼럼 데이터로 매핑
dict = {"col1":[1,11],"col2":[2,22],"col3":[3,33]}
df_dict=pd.DataFrame(dict)
print("딕셔너리로 만든 DataFrame:\n",df_dict)

딕셔너리로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    22    33


## DataFrame -> 넘파이 ndarray, 리스트, 딕셔너리 

#### .values 

In [23]:
#DataFrame => ndarray
array3 = df_dict.values
print("df_dict.values 타입:", type(array3))
print("df_dict.values shape:", array3.shape)
print(array3)

df_dict.values 타입: <class 'numpy.ndarray'>
df_dict.values shape: (2, 3)
[[ 1  2  3]
 [11 22 33]]


In [25]:
#DataFrame => 리스트
list3 = df_dict.values.tolist()
print("df_dict.values.tolist() 타입:", type(list3))
print(list3)

df_dict.values.tolist() 타입: <class 'list'>
[[1, 2, 3], [11, 22, 33]]


DataFrame => 딕셔너리

DataFrame 객체의 to_dict( ) 메서드를 호출하는데, 인자로 'list'를 입력하면 딕셔너리의 값이 리스트형으로 반환

In [26]:
#DataFrame => 딕셔너리
dict3 = df_dict.to_dict("list")
print("df_dict.to_dict()타입:", type(dict3))
print(dict3)

df_dict.to_dict()타입: <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


----------------------------------------------------------------------------------------------------------

## DataFrame의 칼럼 데이터 세트 생성과 수정

In [15]:
titanic_df["Age_0"]=0

In [16]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

In [17]:
titanic_df["Age_by_10"]=titanic_df["Age"]*10
titanic_df["Family_No"]=titanic_df["SibSp"]+titanic_df["Parch"]+1
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_No  
0          2  
1          2  
2          1

In [18]:
titanic_df["Age_by_10"]=titanic_df["Age_by_10"]+100
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

----------------------------------------------------------------------------------------------------------

## DataFrame 삭제

#### drop( ) 메서드

DataFrame.drop(**labels=None**, **axis=0**, index=None, level=None, **inplace=False**,errors="raise")

가장 중요한 파라미터는 labels, axis, inplace

- axis <br>
: 특정 칼럼 또는 특정 행을 드롭.<br>
2차원 ndarray의 구성 시 axis0과 axis1이 있음을 적용 = 판다스의 DataFrame은 2차원 데이터만 다루므로 axis0, axis1<br>
=> axis=1을 입력 ; labels에 원하는 칼럼 명 입력하면 지정된 칼럼을 드롭<br>
=> axis=0을 입력; 특정 로우를 드롭하겠다는 뜻. **DataFrame에서 특정 로우를 가리키는 것은 인덱스** labels에 오는 값을 인덱스로 간주

- inplace<br>
inplace는 디폴트 값이 Fasle이므로 inplace 파라미터를 기재하지않으면 자동으로 False<br>
inplace = False ; 자기 자신의 DataFrame의 데이터는 삭제하지 않으며, 삭제된 결과 DataFrame을 반환<br>
inplace = True ; 자신의 DataFrame의 데이터를 삭제<br>
; 이때 유의할 점은 반환 값이 none이 됨.<br>
따라서 inplace=True로 설정한 채로 반환 값을 다시 자신의 DataFrame 객체로 할당하면 안됨.<br> 
titanic_df= titanic_df.drop(["Age_0","Age_by_10","Family_No"],axis=1,inplace=True)는 titanic_df 객체 변수를 아예 none으로 만들어 버림

또한 여러 개의 칼럼을 삭제하고 싶으면 리스트 형태로 삭제하고자 하는 칼럼 명을 입력해 labels 파라미터로 입력하면 됨. 

In [19]:
titanic_drop_df = titanic_df.drop("Age_0",axis=1)
titanic_drop_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_No  
0      0         A/5 21171   7.2500   NaN        S      320.0          2  
1      0          PC 17599  71.2833   C85        C      480.0          2  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          1

In [20]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

앞의 코드에서 inplace = False가 자동으로 지정되었기 때문에 원본 Titanic DataFrame에는 삭제됐다고 생각했던 "Age_0" 칼럼이 여전히 존재 

In [21]:
drop_result= titanic_df.drop(["Age_0","Age_by_10","Family_No"],axis=1,inplace=True)
print("inplace=True로 drop 후 반환된 값:", drop_result)

inplace=True로 drop 후 반환된 값: None


In [22]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [23]:
pd.set_option("display.width",1000)
pd.set_option("display.max_colwidth",15)
print("#### before axis 0 drop ####")
print(titanic_df.head(3))

titanic_df.drop([0,1,2],axis=0, inplace=True)
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


-------------------------------------------------------------------------------------------------------

## Index 객체

: 판다스의 index 객체는 DataFrame, Series의 레코드를 고유하게 식별하는 객체<br>
: DataFrame, Series에서 Index 객체만 추출 => DataFrame.index / Series.index 속성을 통해 가능<br>
: 반환된 Index 객체의 실제 값은 넘파이 1차원 ndarray로 볼 수 있음<br>
-> Index 객체의 values 속성으로 ndarray 값을 알 수 있다<br>
: ndarray와 유사하게 단일 값 반환 및 슬라이싱도 가능. 하지만 한번 만들어진 DataFrame 및 Series의 index 객체는 함부로 변경할 수 없음

In [24]:
#원본 파일 다시 로딩
titanic_df = pd.read_csv("/Users/ijiseon/Desktop/titanic_train.csv")

In [25]:
#index 객체 추출
indexes = titanic_df.index
print(indexes)

RangeIndex(start=0, stop=891, step=1)


In [27]:
#index 객체를 실제 값 array로 변환
print("index 객체 array 값:\n", indexes.values)

index 객체 array 값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233
 234 235 236 237 238 239 240 241

In [28]:
print(type(indexes.values))
print(indexes.values.shape)

<class 'numpy.ndarray'>
(891,)


In [29]:
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

[0 1 2 3 4]
[0 1 2 3 4]
6


In [30]:
indexes[0]=5

TypeError: Index does not support mutable operations

Series 객체는 Index 객체를 포함하지만,<br>
Series 객체에 연산함수를 적용할 때 Index는 연산에서 제외됩니다.<br>
Indexs는 오직 식별용으로만 사용됩니다. 

In [34]:
series_fair = titanic_df["Fare"]
print("Fair Series max 값:", series_fair.max())
print("Fair Series sum 값:", series_fair.sum())
print("sum() Fair Series:", sum(series_fair))
print("------------------------------------------")
print("Fair Series +3:\n", (series_fair+3).head(3))

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series: 28693.949299999967
------------------------------------------
Fair Series +3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


#### reset_index( ) 메서드 

: 새롭게 인덱스를 연속 숫자 형으로 할당하며 기존 인덱스는 "index"라는 새로운 칼럼 명으로 추가됨<br>
: reset_index()는 인덱스가 연속된 int 숫자형 데이터가 아닐 경우에 다시 이를 연속 int 숫자형 데이터로 만들 때 주로 사용<br>
ex) "Pclass" 칼럼 Series의 value_counts()를 수행하면 "Pclass" 고유 값이 식별자 인덱스 역할을 함<br>
이보다는 연속 숫자형 인덱스가 고유 식별자로 더 적합해보 임 -> "reset_index()" 이용<br>
: Series에 reset_index()를 적용하면, 기존 인덱스가 칼럼으로 추가돼 칼럼이 2개 => DataFrame이 반환<br>
: reset_index()의 parameter 중 drop=True로 설정하면 기존 인덱스는 새로운 칼럼으로 추가되지 않고 삭제(drop)됨<br>
=새로운 칼럼이 추가되지 않으므로 그대로 series로 유지됨 (dataframe으로 반환 X)

In [36]:
titanic_reset_df = titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

In [39]:
print("###Before reset_index###")

value_counts = titanic_df["Pclass"].value_counts()
print(value_counts)
print("value_counts 객체 변수 타입:", type(value_counts))

new_value_counts = value_counts.reset_index(inplace=False)
    
print("###After reset_index###")
print(new_value_counts)
print("new_value_counts 객체 변수 타입:", type(new_value_counts))

###Before reset_index###
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>
###After reset_index###
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


----------------------------------------------------------------------------------------------------

## 데이터 셀렉션 및 필터링

넘파이의 경우 "[ ]" 연산자 내 단일 값 추출, 슬라이싱, 팬시 인덱싱, 불린 인덱싱을 통해 데이터 추출<br>
판다스의 경우 ix[ ], iloc[ ], loc[ ] 연산자를 통해 동일한 작업 수행


**위치 기반 인덱싱과 명칭 기반 인덱싱의 구분**<br>
-위치 기반 인덱싱<br>
: 0을 출발점으로 하는 가로축, 세로축 좌표 기반의 행과 열 위치를 기반으로 데이터를 지정.<br>
: 행, 열 값으로 정수가 입력됨.<br> 
-명칭 기반 인덱싱<br>
: 칼럼의 명칭을 기반으로 위치를 지정하는 방식<br>
: '칼럼 명' 같이 명칭으로 열 위치를 지정하는 방식


### DataFrame의 [ ] 연산자<br>
: 칼럼만 지정할 수 있는 칼럼지정 연산자<br>
넘파이에서 [ ] 연산자는 행의 위치, 열의 위치, 슬라이싱 범위 등을 지정해 데이터를 가져올 수 있음<br>
하지만 DataFrame 바로 뒤에 있는 [ ] 안에 들어갈 수 있는 것은 칼럼 명 문자 또는 칼럼 명의 리스트 객체<br>
: 인덱스로 변환 가능한 표현식<br>
ex) titanic_df의 처음 2개 데이터를 추출하고자 titanic_df[0:2]와 같은 슬라이싱을 이용하면 됨<br>
: 불린 인덱싱 표현도 가능

In [51]:
print("단일 칼럼 데이터 추출:\n", titanic_df["Pclass"].head(3))
print("\n여러 칼럼의 데이터 추출:\n",titanic_df[["Survived","Pclass"]].head(3))
print("[ ]안에 숫자 index는 key error 오류 발생:\n", titanic_df[0])

단일 칼럼 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 칼럼의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3


KeyError: 0

In [52]:
titanic_df[0:2]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked  Name_len Child_Adult Age_cat
0            1         0       3  Braund, Mr....    male  22.0      1      0  A/5 21171   7.2500   NaN        S        23       Adult   Adult
1            2         1       1  Cumings, Mr...  female  38.0      1      0   PC 17599  71.2833   C85        C        51       Adult   Adult

In [53]:
titanic_df[titanic_df["Pclass"]==3].head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket   Fare Cabin Embarked  Name_len Child_Adult Age_cat
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171  7.250   NaN        S        23       Adult   Adult
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...  7.925   NaN        S        22       Adult   Adult
4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450  8.050   NaN        S        24       Adult   Adult

- DataFrame ix[ ] 연산자 ; 향후 사라질 예정

In [54]:
print("칼럼 위치 기반 인덱싱 데이터 추출:", titanic_df.ix[0,2])
print("칼럼 명 기반 인덱싱 데이터 추출:", titanic_df.ix[0,"Pclass"])

AttributeError: 'DataFrame' object has no attribute 'ix'

In [56]:
data = {"Name" : ["Chulmin","Eunkyung","Jinwoong","Soobeom"],
       "Year": [2011,2016,2015,2015],
       "Gender":["Male","Female","Male","Male"]}
data_df = pd.DataFrame(data, index=["one","two","three","four"])
data_df

Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

In [59]:
#data_df를 reset_index()로 새로운 숫자형 인덱스를 생성
data_df_reset = data_df.reset_index()
data_df_reset = data_df_reset.rename(columns={"index":"old_index"})

#인덱스값에 1을 더해서 1부터 시작하는 새로운 인덱스 값 생성
data_df_reset.index = data_df_reset.index+1
data_df_reset

old_index      Name  Year  Gender
1       one   Chulmin  2011    Male
2       two  Eunkyung  2016  Female
3     three  Jinwoong  2015    Male
4      four   Soobeom  2015    Male

### DataFrame iloc[ ] 연산자<br>
: 위치 기반 인덱싱<br>
=위치 인덱싱이 아닌 명칭을 입력하면 오류가 발생<br>
: 행과 열 값으로 integer 또는 integer형의 슬라이싱, 팬시 리스트 값을 입력해줘야함.<br>
: iloc[ ]은 슬라이싱과 팬시 인덱싱은 제공하나 명확한 위치 기반 인덱싱이 사용되어야 하는 제약으로 인해 불린 인덱싱은 제공하지 않음<br>

In [60]:
data_df.iloc[0,0]

'Chulmin'

In [61]:
#위치 인덱싱이 아닌 명칭을 입력하면 오류 발생
data_df.iloc[0,"Name"]

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

In [63]:
#위치인덱싱=> 문자열 인덱스를 행 위치에 입력해도 오류가 발생
data_df.iloc["one",0]

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

### DataFrame loc[ ] 연산자<br>
: 명칭 기반 인덱싱<br>
: 행 위치에는 DataFrame inde 값을, 그리고 열 위치에는 칼럼명 입력<br>
: index가 숫자 형일 수 있기 때문에 명칭 기반이라고 무조건 문자열을 입력한다는 선입견을 가져서는 안됨<br>
: 슬라이싱할 때 주의할 점 <br>
일반적으로 슬라이싱을 '시작값:종료값'과 같이 지정하면 시작 값 ~ 종료값-1<br>
그런데 loc[ ]에 슬라이싱 기호를 적용하면 종료값-1 이 아니라 **종료값까지 포함**<br>
이는 명칭 기반 인덱싱의 특성 때문 => 명칭은 숫자형이 아닐 수 있기 때문에 -1을 할 수가 없음<br>

In [64]:
data_df.loc["one","Name"] #인덱스값이 "one"인 행의 칼럼 명이 "name"인 데이터 

'Chulmin'

In [67]:
data_df_reset.loc[1,"Name"]

'Chulmin'

In [69]:
data_df_reset.loc[0,"Name"] #data_df_reset에서는 index가 0인 행이 없기 때문에 오류 반환

KeyError: 0

In [70]:
print("위치 기반 iloc slicing\n", data_df.iloc[0:1,0], "\n")
print("명칭 기반 loc slicing\n", data_df.loc["one":"two","Name"])

위치 기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 

명칭 기반 loc slicing
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object


In [72]:
print(data_df_reset.loc[1:2, "Name"])

1     Chulmin
2    Eunkyung
Name: Name, dtype: object


### 불린인덱싱 

In [74]:
titanic_df = pd.read_csv("/Users/ijiseon/Desktop/titanic_train.csv")

In [75]:
titanic_boolean = titanic_df[titanic_df["Age"]>60]
print(type(titanic_boolean))
titanic_boolean

<class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked
33            34         0       2  Wheadon, Mr...    male  66.0      0      0   C.A. 24579   10.5000          NaN        S
54            55         0       1  Ostby, Mr. ...    male  65.0      0      1       113509   61.9792          B30        C
96            97         0       1  Goldschmidt...    male  71.0      0      0     PC 17754   34.6542           A5        C
116          117         0       3  Connors, Mr...    male  70.5      0      0       370369    7.7500          NaN        Q
170          171         0       1  Van der hoe...    male  61.0      0      0       111240   33.5000          B19        S
252          253         0       1  Stead, Mr. ...    male  62.0      0      0       113514   26.5500          C87        S
275          276         1       1  Andrews, Mi...  female  63.0      1      0        13502   77.9583           D7        S
280          281         0       3  Duane, Mr. ...    male  65.0      0      0       336439    7.7500          NaN        Q
326          327         0       3  Nysveen, Mr...    male  61.0      0      0       345364    6.2375          NaN        S
438          439         0       1  Fortune, Mr...    male  64.0      1      4        19950  263.0000  C23 C25 C27        S
456          457         0       1  Millet, Mr....    male  65.0      0      0        13509   26.5500          E38        S
483          484         1       3  Turkula, Mr...  female  63.0      0      0         4134    9.5875          NaN        S
493          494         0       1  Artagaveyti...    male  71.0      0      0     PC 17609   49.5042          NaN        C
545          546         0       1  Nicholson, ...    male  64.0      0      0          693   26.0000          NaN        S
555          556         0       1  Wright, Mr....    male  62.0      0      0       113807   26.5500          NaN        S
570          571         1       2  Harris, Mr....    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S
625          626         0       1  Sutton, Mr....    male  61.0      0      0        36963   32.3208          D50        S
630          631         1       1  Barkworth, ...    male  80.0      0      0        27042   30.0000          A23        S
672          673         0       2  Mitchell, M...    male  70.0      0      0   C.A. 24580   10.5000          NaN        S
745          746         0       1  Crosby, Cap...    male  70.0      1      1    WE/P 5735   71.0000          B22        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0       113572   80.0000          B28      NaN
851          852         0       3  Svensson, M...    male  74.0      0      0       347060    7.7750          NaN        S

In [76]:
titanic_df[titanic_df["Age"]>60][["Name","Age"]].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [78]:
titanic_df.loc[titanic_df["Age"]>60,["Name","Age"]].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [79]:
titanic_df[(titanic_df["Age"]>60)&(titanic_df["Pclass"]==1)&(titanic_df["Sex"]=="female")]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

In [80]:
cond1 =titanic_df["Age"]>60
cond2 =titanic_df["Pclass"]==1
cond3 =titanic_df["Sex"]=="female"
titanic_df[cond1&cond2&cond3]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

------------------------------------------------------------------------------------------------------------------------------------------------------------

##  정렬, Aggregation 함수, GroupBy 적용

### DataFrame, Series의 정렬 - sort_values( )

sort_values()의 주요 입력 파라미터 - by, ascending, inplace

- by<br>
: 특정 칼럼을 입력하면 해당 칼럼으로 정렬을 수행<br>
: 여러 개의 칼럼으로 정렬하려면 by에 리스트 형식으로 정렬하는 칼럼을 입력하면 됨<br> 
추가) 리스트 형식으로 정렬할 때 칼럼별 순서에 따라 정렬 정리됨 (=칼럼입력순위가 정렬우선순위)<br>
예를 들어 by = [칼럼1,칼럼2] 면 칼럼1 정렬에 따른 칼럼2 정렬<br> 
- ascending<br>
: ascending =True는 오름차순 / ascending =False는 내림차순<br>
- inplace<br>
: inplace = False로 설정하면, sort_values()를 호출한 DataFrame은 그대로 유지하며, 정렬된 DataFrame을 결과로 반환<br>
: inplace = True로 설정하면, 호출한 DataFrame의 정렬 결과를 그대로 적용. 기본은 inplace=False

In [81]:
titanic_sorted = titanic_df.sort_values(by=["Name"])
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket   Fare Cabin Embarked
845          846         0       3  Abbing, Mr....    male  42.0      0      0  C.A. 5547   7.55   NaN        S
746          747         0       3  Abbott, Mr....    male  16.0      1      1  C.A. 2673  20.25   NaN        S
279          280         1       3  Abbott, Mrs...  female  35.0      1      1  C.A. 2673  20.25   NaN        S

In [82]:
titanic_sorted = titanic_df.sort_values(by=["Pclass","Name"],ascending=False)
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name   Sex   Age  SibSp  Parch    Ticket  Fare Cabin Embarked
868          869         0       3  van Melkebe...  male   NaN      0      0    345777   9.5   NaN        S
153          154         0       3  van Billiar...  male  40.5      0      2  A/5. 851  14.5   NaN        S
282          283         0       3  de Pelsmaek...  male  16.0      0      0    345778   9.5   NaN        S

In [87]:
#추가코드
titanic_sorted = titanic_df.sort_values(by=["Name","Pclass"])
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket   Fare Cabin Embarked
845          846         0       3  Abbing, Mr....    male  42.0      0      0  C.A. 5547   7.55   NaN        S
746          747         0       3  Abbott, Mr....    male  16.0      1      1  C.A. 2673  20.25   NaN        S
279          280         1       3  Abbott, Mrs...  female  35.0      1      1  C.A. 2673  20.25   NaN        S

In [90]:
#추가코드
titanic_sorted = titanic_df.sort_values(by=["Pclass","Name"])
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name     Sex    Age  SibSp  Parch  Ticket      Fare    Cabin Embarked
730          731         1       1  Allen, Miss...  female  29.00      0      0   24160  211.3375       B5        S
305          306         1       1  Allison, Ma...    male   0.92      1      2  113781  151.5500  C22 C26        S
297          298         0       1  Allison, Mi...  female   2.00      1      2  113781  151.5500  C22 C26        S

In [91]:
#추가코드
titanic_sorted = titanic_df.sort_values(by=["Name","Pclass"], ascending=[True,False])
titanic_sorted

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch         Ticket     Fare Cabin Embarked
845          846         0       3  Abbing, Mr....    male  42.0      0      0      C.A. 5547   7.5500   NaN        S
746          747         0       3  Abbott, Mr....    male  16.0      1      1      C.A. 2673  20.2500   NaN        S
279          280         1       3  Abbott, Mrs...  female  35.0      1      1      C.A. 2673  20.2500   NaN        S
308          309         0       2  Abelson, Mr...    male  30.0      1      0      P/PP 3381  24.0000   NaN        C
874          875         1       2  Abelson, Mr...  female  28.0      1      0      P/PP 3381  24.0000   NaN        C
..           ...       ...     ...             ...     ...   ...    ...    ...            ...      ...   ...      ...
286          287         1       3  de Mulder, ...    male  30.0      0      0         345774   9.5000   NaN        S
282          283         0       3  de Pelsmaek...    male  16.0      0      0         345778   9.5000   NaN        S
361          362         0       2  del Carlo, ...    male  29.0      1      0  SC/PARIS 2167  27.7208   NaN        C
153          154         0       3  van Billiar...    male  40.5      0      2       A/5. 851  14.5000   NaN        S
868          869         0       3  van Melkebe...    male   NaN      0      0         345777   9.5000   NaN        S

[891 rows x 12 columns]

### Aggregation 함수 적용 
#### DataFrame에서 min(), max(), sum(), count()와 같은 aggregation 함수의 적용 
DataFrame에서 바로 aggregation을 호출할 경우 **모든 칼럼**에 해당 aggregation을 적용<br>
특정 칼럼에 aggregation 함수를 적용하기 위해서는 DataFrame에 대상 칼럼들만 추출해 aggregation을 적용

In [92]:
titanic_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [93]:
titanic_df[["Age","Fare"]].mean()

Age     29.699118
Fare    32.204208
dtype: float64

### groupby( ) 적용

DataFrame의 groupby() 사용 시 입력 파라미터 by에 칼럼을 입력하면 대상 칼럼으로 groupby됨<br>
DataFrame에 groupby()를 호출하면 DataFrameGroupBy라는 또 다른 형태의 DataFrame을 반환<br>
-> 그리고 그 반환된 결과에 aggregation 함수를 호출하면 groupby()대상 칼럼을 제외한 모든 칼럼에 해당 aggregation 함수를 적용 

In [97]:
titanic_groupby = titanic_df.groupby(by="Pclass")
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


In [100]:
titanic_groupby = titanic_df.groupby(by="Pclass").count()
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  Cabin  Embarked
Pclass                                                                                    
1               216       216   216  216  186    216    216     216   216    176       214
2               184       184   184  184  173    184    184     184   184     16       184
3               491       491   491  491  355    491    491     491   491     12       491

In [103]:
#추가코드 
titanic_groupby1 = titanic_df.groupby(by=["Pclass","Sex"]).count()
titanic_groupby1

PassengerId  Survived  Name  Age  SibSp  Parch  Ticket  Fare  Cabin  Embarked
Pclass Sex                                                                                  
1      female           94        94    94   85     94     94      94    94     81        92
       male            122       122   122  101    122    122     122   122     95       122
2      female           76        76    76   74     76     76      76    76     10        76
       male            108       108   108   99    108    108     108   108      6       108
3      female          144       144   144  102    144    144     144   144      6       144
       male            347       347   347  253    347    347     347   347      6       347

In [102]:
titanic_groupby = titanic_df.groupby(by="Pclass")[["PassengerId","Survived"]].count()
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

In [104]:
titanic_df.groupby("Pclass")["Age"].agg([max,min])

max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

In [106]:
agg_format = {"Age":"max","SibSp":"sum","Fare":"mean"}
titanic_df.groupby("Pclass").agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

----------------------------------------------------------------------------------------------------

## 결손 데이터 처리하기 

결손데이터(Missing Data)<br>
결손 데이터는 칼럼에 값이 없는, 즉 NULL인 경우를 의미, 이를 넘파이의 NaN으로 표시

기본적으로 머신러닝 알고리즘은 이 NaN 값을 처리하지 않으므로 이 값을 다른 값으로 대체해야 함.<br>
또한 NaN 값은 평균, 총합 등의 함수 연산 시 제외 가 됨<br>
ex) 특정 칼럼의 100개 데이터 중 10개가 NaN 값일 경우, 이 칼럼의 평균 값은 90개 데이터에 대한 평균

isna(): NaN 여부를 확인하는 API<br>
fillna(): NaN 값을 다른 값으로 대체하는 API

#### isna( )로 결손 데이터 여부 확인

결손 데이터의 개수는 isna( )결과에 sum( )함수를 추가해 구할 수 있음<br>
=> DataFrame에 isna()를 수행하면 모든 칼럼의 값이 NaN인지 아닌지를 True나 False로 알려줌<br>
=> 여기에 sum()을 호출 시 True는 내부적으로 숫자 1로, False는 숫자 0으로 변환됨 ; 결손 데이터의 개수를 구할 수 있음

In [107]:
titanic_df.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  False  False  False  False   False  False   True     False
1        False     False   False  False  False  False  False  False   False  False  False     False
2        False     False   False  False  False  False  False  False   False  False   True     False

In [108]:
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

#### fillna( )로 결손 데이터 대체하기 

주의해야 할 점은 fillna()를 이용해 반환 값을 다시 받거나 inplace=True 파라미터를 fillna()에 추가해야 실제 데이터 세트값이 변경됨<br>
ex1) titanic_df["Cabin"]=titanic_df["Cabin"].fillna("C000")<br>
ex2) titanic_df["Cabin"].fillna("C000",inplace=True)

In [109]:
titanic_df["Cabin"]=titanic_df["Cabin"].fillna("C000")
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500  C000        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250  C000        S

In [110]:
titanic_df["Age"]=titanic_df["Age"].fillna(titanic_df["Age"].mean())
titanic_df["Embarked"]=titanic_df["Embarked"].fillna("S")
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

-------------------------------------------------------------------------------------

## apply lambda 식으로 데이터 가공

#### lambda

lambda x : x^2에서 " : " 로 입력 인자와 반환될 입력 인자의 계산식을 분리<br>
":" 왼쪽에 있는 x는 입력 인자 / 오른쪽은 입력인자의 계산 식 즉 반환 값을 의미<br>

여러 개의 값을 입력인자로 사용해야 할 경우에는 보통 map()함수를 결합해서 사용

In [40]:
def get_square(a):
    return a**2

print("3의 제곱은:", get_square(3))

3의 제곱은: 9


In [41]:
lambda_square = lambda x : x**2

print("3의 제곱은:", lambda_square(3))

3의 제곱은: 9


In [44]:
#입력인자가 여러개 
a = [1,2,3]
squares = map(lambda x : x**2, a)
list(squares)

[1, 4, 9]

In [45]:
map(lambda x : x**2, a)

### 판다스 DataFrmae의 lambda 식 

lambda 식은 if else를 지원<br>
주의할 점_1: if 절의 경우 if 식보다 반환 값을 먼저 기술해야 함.<br>
왜냐하면 lambda 식 ":" 기호의 오른편에 반환 값이 있어야 하기 때문. else의 경우는 else 식이 먼저 나오고 반환 값이 나중에 오면 됨<br>
주의할 점_2: if, else만 지원하고, if-else if-else는 지원하지 않음.<br>
else 절을 ()로 내포해 () 내에서 다시 if else를 적용해 사용

In [111]:
titanic_df["Name_len"] = titanic_df["Name"].apply(lambda x: len(x))
titanic_df[["Name","Name_len"]].head(3)

Name  Name_len
0  Braund, Mr....        23
1  Cumings, Mr...        51
2  Heikkinen, ...        22

In [112]:
#나이가 15세 미만이면 "Child", 그렇지 않으면 "Adult"로 구분하는 새로운 칼럼 "Child_Adult"

titanic_df["Child_Adult"] = titanic_df["Age"].apply(lambda x : "Child" if x <=15 else "Adult")
titanic_df[["Age","Child_Adult"]].head(8)

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

In [113]:
#나이가 15세 미만이면 "Child", 15세에서 60세는 "Adult", 61세 이상은 "Elderly" 구분하는 새로운 칼럼 "Age_cat"

titanic_df["Age_cat"]=titanic_df["Age"].apply(lambda x: "Child" if x<=15
                                              else ("Adult" if x<=60 else "Elderly"))

titanic_df["Age_cat"].value_counts()

Adult      786
Child       83
Elderly     22
Name: Age_cat, dtype: int64

else if가 많이 나와야 하는 경우나 switch case문의 경우는 이렇게 else를 계속 내포해서 쓰기 부담<br>
-> 별도의 함수를 만드는 게 더 나음 

In [114]:
#나이에 따라 세분화된 분류를 수행하는 함수 생성
def get_category(age):
    cat=""
    if age <=5: cat="Baby"
    elif age <=12: cat="Child"
    elif age <=18: cat="Teenager"
    elif age <=25: cat="Student"
    elif age <=35: cat="Young Adult"
    elif age <=60: cat="Adult"
    else: cat="Elderly"
        
    return cat

In [115]:
#lambda식에 위에서 생성한 get_category()함수를 반환값으로 지정
#get_category(x)는 입력값으로 "Age" 칼럼 값을 받아서 해당하는 cat을 반환

titanic_df["Age_cat"]=titanic_df["Age"].apply(lambda x : get_category(x))
titanic_df[["Age","Age_cat"]].head()

Age      Age_cat
0  22.0      Student
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult